File: 01-Programming-Languages/Java/Algorithms/Graph/BFS.java
/*
 * Problem: Traverse a graph using Breadth-First Search (BFS).
 * Approach: Use a queue to explore neighbors level by level.
 * Time Complexity: O(V + E)
 * Space Complexity: O(V)
 * Example: Graph: 0-1, 0-2, 1-3, 1-4, 2-5, 2-6 -> BFS from 0: 0 1 2 3 4 5 6
 */

import java.util.*;

public class BFS {
    public static List<Integer> bfs(Map<Integer, List<Integer>> graph, int start) {
        List<Integer> result = new ArrayList<>();
        Queue<Integer> queue = new LinkedList<>();
        Set<Integer> visited = new HashSet<>();
        queue.offer(start);
        visited.add(start);
        while (!queue.isEmpty()) {
            int node = queue.poll();
            result.add(node);
            for (int neighbor : graph.getOrDefault(node, Collections.emptyList())) {
                if (!visited.contains(neighbor)) {
                    visited.add(neighbor);
                    queue.offer(neighbor);
                }
            }
        }
        return result;
    }

    public static void main(String[] args) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        graph.put(0, Arrays.asList(1, 2));
        graph.put(1, Arrays.asList(3, 4));
        graph.put(2, Arrays.asList(5, 6));
        System.out.println(bfs(graph, 0));
    }
}

File: 01-Programming-Languages/Kotlin/Algorithms/Graph/bfs.kt
/*
 * Problem: Traverse a graph using Breadth-First Search (BFS).
 * Approach: Use a queue to explore neighbors level by level.
 * Time Complexity: O(V + E)
 * Space Complexity: O(V)
 * Example: Graph: 0-1, 0-2, 1-3, 1-4, 2-5, 2-6 -> BFS from 0: [0, 1, 2, 3, 4, 5, 6]
 */

import java.util.*

fun bfs(graph: Map<Int, List<Int>>, start: Int): List<Int> {
    val result = mutableListOf<Int>()
    val queue: Queue<Int> = LinkedList()
    val visited = mutableSetOf<Int>()
    queue.offer(start)
    visited.add(start)
    while (queue.isNotEmpty()) {
        val node = queue.poll()
        result.add(node)
        for (neighbor in graph[node].orEmpty()) {
            if (neighbor !in visited) {
                visited.add(neighbor)
                queue.offer(neighbor)
            }
        }
    }
    return result
}

fun main() {
    val graph = mapOf(
        0 to listOf(1, 2),
        1 to listOf(3, 4),
        2 to listOf(5, 6)
    )
    println(bfs(graph, 0))
}
